### ISSUE TITLE
[HOLD for Payment 8-14-2023][$2000]  Text is not centered in codeblock on android app

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18913]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app in android
2. Open any report
3. send 3 backtick codeblock, to observe it clearly, send 3 backtick message with numbers on new line like shown in video
4. Observe that text inside codeblock is not centered

## Expected Result: 
App should center the text in codeblocks
## Actual Result: 
Codeblock text is not centered on android app
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.13.3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/c289c601-e278-49ee-b6e6-f22e03a33b0a

<img src="https://github.com/Expensify/App/assets/43996225/db36bc18-9616-4b9d-858d-4cec0793d1f1" width="40%" height="40%">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684053742783289

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018c9148c105260d97</li>
        <li>Upwork Job ID: 1659251229090435072</li>
        <li>Last Price Increase: 2023-06-07</li>
    </ul>
</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
A multiline text inside code block is not centered in android.

### What is the root cause of that problem?
This is a bug in React Native Text component (not reproducible in native Android). It is noticeable on code block because it's the only markdown that allow multiline with background color. In Android, when we have a new line, the text width will be different with the one that don't have a newline. Below is the image comparison:
<img width="798" alt="image" src="https://github.com/Expensify/App/assets/50919443/8573d2c5-1110-4a17-b457-2d4daba73f68">

### What changes do you think we should make in order to solve the problem?


I don't found any existing issue reported in RN repo. In the meantime, we can align the text to the center by adding `textAlign` center to the `pre` style.

https://github.com/Expensify/App/blob/83d9ac3729dfc8c8804624c8aec7113c7ffb9f8e/src/styles/styles.js#L106-L115

This will align the text to center but the width is still bigger than the text width no newline.

**Comment by adelekennedy** [comment]
asked clarifying question in Slack

**Comment by s77rt** [comment]
@bernhardoj Thanks for the proposal. If the bug is coming from RN then we should probably fix it upstream.

**Comment by adelekennedy** [comment]
@Christinadobrzyn I'm ooo until Monday - I'm adding you here to babysit and then I will assign back to myself!

**Comment by Uzayn** [comment]
## Proposal

**Problem Description:**
The issue reported is regarding the text alignment within codeblocks in the Expensify Android app. Currently, the text inside codeblocks is not centered, which affects the overall readability and user experience.

**Reproduction Steps:**
1. Open the Expensify Android app.
2. Navigate to any report.
3. Observe the codeblocks and notice that the text inside them is not centered.

**Proposed Solution:**
To address this issue, I propose implementing the following solution:
1. Modify the `WrappedText` component, which handles the rendering of codeblocks.
2. Update the `codeWordWrapper` style in the `codeStyles.js` file to include `alignItems: 'center'`. This will horizontally center the text within each codeblock.

**Implementation Plan:**
1. Locate the `WrappedText` component in the codebase.
2. Find the `codeWordWrapper` style definition in the `codeStyles.js` file.
3. Update the `codeWordWrapper` style to include `alignItems: 'center'`.
4. Rebuild and test the Android app to ensure the text alignment within codeblocks is now centered.

**Testing and Validation:**
To ensure the proposed solution is robust and meets the requirements:
1. Write test cases to cover different scenarios of codeblocks.
2. Execute the test cases on the Android platform.
3. Validate that the text inside codeblocks is centered correctly.
4. Perform additional UI testing to ensure the overall app functionality is not affected.

**Timeline:**
The estimated timeline for implementing this solution is as follows:
- Research and code modifications: 2 days
- Testing and validation: 1 day
- Bug fixes and refinements: 1 day

**Platform Screenshots:**

Screenshot of the iOS code with justify center property:
![Screenshot (1023)](https://github.com/Expensify/App/assets/95490624/65887692-6583-458b-8e2d-3af4eceb395f)



Screenshot of the Android code without justify center property:
![Screenshot (1022)](https://github.com/Expensify/App/assets/95490624/db844725-64f5-4b88-a757-94594af62d24)

**Comment by Uzayn** [comment]
Contributor details
Your Expensify account email: koleoshohuzain@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01eb5b8faf0bbe0655

**Comment by s77rt** [comment]
@Uzayn Thanks for the proposal. I'm not sure you understood the problem. We are not trying to centre every line horizontally. We are looking to have the padding/width in Android consist with other platforms.

| :x:  | :heavy_check_mark:  |
|--------|-------|
|    <img src='https://github.com/Expensify/App/assets/16493223/e99adecc-5f76-4507-8947-6e0fd592c158' />    |   <img src='https://github.com/Expensify/App/assets/16493223/d38874e2-6b6d-471b-8f4f-86535155d069' />    |

**Comment by s77rt** [comment]
@Uzayn Thanks for the quick follow up. I don't think the font has anything to do here, even if the letters are not taking the same space there should not be an empty space at the end of each line.

(We do use a monospaced font already: ExpensifyMono)

**Comment by bernhardoj** [comment]
Open an upstream PR [here](https://github.com/facebook/react-native/pull/37590). Added optional solution to solve the issue upstream.

**Comment by s77rt** [comment]
@bernhardoj Thanks! I have a question, if you type `1    \n` will the code block look as expected i.e with trailing white spaces? The Android doc is not clear if all the trailing white-space will be removed or just the newline one.

**Comment by bernhardoj** [comment]
That's actually a good question. From my test, it will still includes the space when we have a new line character.

<img width="532" alt="image" src="https://github.com/Expensify/App/assets/50919443/1b86879a-0911-4f9e-a491-23bfde566427">

But it won't include the whitespace if the line does not contains new line character, which becomes an issue. I have an explanation [here](https://github.com/facebook/react-native/pull/37590#issuecomment-1565209279) too.

**Comment by s77rt** [comment]
@bernhardoj Thanks for the quick response. Let's see how it goes on RN PR.

**Comment by s77rt** [comment]
We are still working on this. This is a bug in RN. We have a possible solution but still being evaluated.

**Comment by Christinadobrzyn** [comment]
Feel free to let me know if you'd like me to raise the price to see more proposals @s77rt!

**Comment by s77rt** [comment]
@Christinadobrzyn The current price is alright no need for a raise here.

**Comment by s77rt** [comment]
We have an upstream PR https://github.com/facebook/react-native/pull/37590. Should be close getting this fixed.

**Comment by s77rt** [comment]
Getting close. We have an approval on the upstream PR https://github.com/facebook/react-native/pull/37590. Hopefully it will get merged soon.

@alex-mechler Can you please assign @bernhardoj 

@bernhardoj Can you please raise the same PR in our fork once assigned

**Comment by s77rt** [comment]
@bernhardoj Based on this slack thread https://expensify.slack.com/archives/C01GTK53T8Q/p1686159512184629 we may not want a PR in our (to be retired) fork. So let's just wait for RN PR I suppose.

**Comment by bernhardoj** [comment]
Got it!

**Comment by Christinadobrzyn** [comment]
Invited to Upwork - https://www.upwork.com/ab/applicants/1659251229090435072/job-details

@s77rt as C+
@bernhardoj as contributor
@dhanashree-sawant as reporter

**Comment by dhanashree-sawant** [comment]
Thanks @Christinadobrzyn, I have accepted the offer.

**Comment by alex-mechler** [comment]
@bernhardoj what is the updated PR link? I see that the one you posted was closed and reverted due to regressions

**Comment by bernhardoj** [comment]
@alex-mechler Sorry, I totally forget to give an update here.

Yes, the previous one is reverted and the new one is [here](https://www.github.com/facebook/react-native/pull/37790) (merged)

**Comment by Christinadobrzyn** [comment]
I think this can be moved to weekly while we review the PR.

**Comment by Christinadobrzyn** [comment]
hey @s77rt sorry, a little lost on this Gh - can you confirm where we are and if there's a PR to follow?

**Comment by s77rt** [comment]
@Christinadobrzyn The PR https://github.com/facebook/react-native/pull/37790 is merged. We need to wait for the next RN release that would include this fix. Usually we would find some ongoing discussions here https://github.com/reactwg/react-native-releases/discussions on what commits to cherry pick for the next release (cc @bernhardoj).

I think we should put this on hold for updating RN. This will likely take some time

**Comment by bernhardoj** [comment]
I checked our commit and it hasn't been picked up yet on any release. Do we need to comment on those discussions to CP the commit? If yes, how do we decide which version we want our commit to be CP-ed?

_If we don't CP, I think the earliest version on which our commit will be included is in 0.73.0_

**Comment by s77rt** [comment]
@bernhardoj 

> Do we need to comment on those discussions to CP the commit? If yes, how do we decide which version we want our commit to be CP-ed?


Yes, the 0.72.1 but not sure if that commit is eligible for that early release. If it's not we will probably have to move this to Monthly.

**Comment by bernhardoj** [comment]
The fix doesn't sound critical, but I [commented](https://www.github.com/reactwg/react-native-releases/discussions/74#discussioncomment-6311375) to CP it. Let's see if it gets picked or not.

**Comment by bernhardoj** [comment]
It gets picked 🎉. We just need to wait our App to update the RN version to 0.72.1. I see we have a PR to [update RN to 0.72](https://github.com/Expensify/App/pull/18507), it should be easy to update it later to 0.72.1.

**Comment by MonilBhavsar** [comment]
👍 Updated title to put it on HOLD

**Comment by Christinadobrzyn** [comment]
waiting on - https://github.com/Expensify/App/pull/18507

**Comment by Christinadobrzyn** [comment]
waiting on - https://github.com/Expensify/App/pull/18507

**Comment by Christinadobrzyn** [comment]
I'm going to be ooo until August 31st so going to unassign and assign a new teammate. 

At this time, we're waiting on https://github.com/Expensify/App/pull/18507

I'll take this back if it's still open when I return

**Comment by MonilBhavsar** [comment]
still on hold and bumping to weekly

**Comment by lschurr** [comment]
On hold.

**Comment by Christinadobrzyn** [comment]
I'm back from ooo - @lschurr thanks for handling this! I can take it off your plate!

**Comment by s77rt** [comment]
@Christinadobrzyn This is fixed now! How can we tell Melvin to hold this on payment? Or have to be done manually?

**Comment by Christinadobrzyn** [comment]
Awesome! @s77rt I think the payments need to manually be done. Payment is 8-14. 

Is this accurate?

Payouts due:

Issue Reporter: $250 @dhanashree-sawant
Contributor: $2000 @bernhardoj 
Contributor+: $2000 @s77rt 

Eligible for 50% #urgency bonus? N

Upwork job is https://www.upwork.com/jobs/~018c9148c105260d97

**Comment by Christinadobrzyn** [comment]
Do we need a regression test?

**Comment by s77rt** [comment]
@Christinadobrzyn Looks accurate!

Given this bug was upstream we won't need a regression test.

**Comment by MonilBhavsar** [comment]
Bumping to weekly, as it is on hold till 14/8

**Comment by Christinadobrzyn** [comment]
Sorry - late to pay this - paying now based on - https://github.com/Expensify/App/issues/18913#issuecomment-1668595526

**Comment by s77rt** [comment]
@Christinadobrzyn Yes. Thank you!

**Comment by MonilBhavsar** [comment]
Issuing payments here

**Comment by Christinadobrzyn** [comment]
Awesome! Everyone has been paid through Upwork - Upwork job is closed - closing this GH.
